File: Py-Solutions/Problem-14.py
# Which starting number, under one million, produces the longest Collatz chain?

def longestCollatz(limit):
  max_len = 0
  max_starter = 0
  for n in range(3,limit):
    starter = n
    chain_len = 1
    while n > 1:
      if n % 2 == 0:
        n = n/2
      else:
        n = 3*n+1
      chain_len += 1
    if chain_len > max_len:
      max_starter = starter
      max_len = chain_len
  return([max_starter, max_len])

def main():
  print(longestCollatz(1000000))

if __name__ == '__main__':
  main()

File: C++Solutions/Problem-14.cpp
/*
Which starting number, under one thousand, produces the longest Collatz chain?
*/

#include <iostream>
using namespace std;

int collatzLength(int num);
int longestCollatz(int limit);

int main() {
  int answer = longestCollatz(1000);
  cout << answer << "\n";
  return 0;
}

int collatzLength(int num) {
  int length = 1;
  while (num > 1) {
    if (num % 2 == 0) {
      num = num / 2;
    } else {
      num = (num*3) + 1;
    }
    length += 1;
  }
  return length;
}

int longestCollatz(int limit) {
  int longestNum = 0, longestChain = 0, i = 2;
  int testLength;
  while (i < limit) {
    testLength = collatzLength(i);
    if (testLength > longestChain) {
      longestNum = i;
      longestChain = testLength;
    }
    i += 1;
  }
  return longestNum;
}
